pkg/ruler/api.go

```diff
@@ -0,0 +1,264 @@
+package ruler
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gorilla/mux"
+	"github.com/weaveworks/common/user"
+)
+
+// RegisterRoutes registers the configs API HTTP routes with the provided Router.
+func (r *Ruler) RegisterRoutes(router *mux.Router) {
+	if r.store == nil {
+		level.Info(util.Logger).Log("msg", "ruler configured with store that does not support api")
+		return
+	}
+	for _, route := range []struct {
+		name, method, path string
+		handler            http.HandlerFunc
+	}{
+		{"get_rules", "GET", "/api/prom/api/v1/rules", r.rules},
+		{"get_alerts", "GET", "/api/prom/api/v1/rules", r.alerts},
+	} {
+		level.Debug(util.Logger).Log("msg", "ruler: registering route", "name", route.name, "method", route.method, "path", route.path)
+		router.Handle(route.path, route.handler).Methods(route.method).Name(route.name)
+	}
+}
+
+// In order to reimplent the prometheus rules API, a large amount of code was copied over
+// This is required because the prometheus api implementation does not pass a context to
+// the rule retrieval function.
+// https://github.com/prometheus/prometheus/blob/2aacd807b3ec6ddd90ae55f3a42f4cffed561ea9/web/api/v1/api.go#L108
+// https://github.com/prometheus/prometheus/pull/4999
```

Comment:
As a non-blocking comment, would be great to double check with @tomwilkie if there's anything we can do to avoid copy-pasting so much code here and reconsider the PR https://github.com/prometheus/prometheus/pull/4999

Reply:
This would go a long way in running cortex as a single binary.